api/src/main/java/io/strimzi/api/kafka/model/ListenersStatus.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2019, Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.api.kafka.model;
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonPropertyOrder;
+import io.strimzi.crdgenerator.annotations.Description;
+import io.sundr.builder.annotations.Buildable;
+import lombok.EqualsAndHashCode;
+
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.Map;
+
+import static java.util.Collections.emptyMap;
+
+@Buildable(
+        editableEnabled = false,
+        generateBuilderPackage = false,
+        builderPackage = "io.fabric8.kubernetes.api.builder"
+)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonPropertyOrder({ "plain", "tls", "external"})
+@EqualsAndHashCode
+public class ListenersStatus implements UnknownPropertyPreserving, Serializable {
```

Comment:
Do we support having multiple listeners of the same type? If not, this change would be for keeping the status flexible so that it could support multiple listeners of the same type when we get there, right?

Reply:
We currently don't, but some people are interesting in having multiple external listeners. Which raises the question of how that should look in the CR. Right now the property name implies the type, but having a property called `external2` is pretty lame. So we _might_ end up using a list of listeners. Which would look somewhat like the above with a list of objects with a name/type and the rest of the listener config.

_If_ that's where we're headed then not making the mistake here of using an object with particular properties would mean making that change for the spec would be less noisy.

But first we need to find a consensus about the CRD changes we would make for supporting multiple external listeners.

Reply:
@tombentley +1

Reply:
The problem is that lists are hard to deal with. Having list objects is much easier to find your particular address.

Reply:
@scholzj I know but how do you propose multiple external listeners would be expressed in a CR, bearing in mind that the kube recommendation is for lists of objects rather than objects being used as maps. Or do you advocate sticking with an object and using a property like `external2`? I only proposed using a list of objects here because I think a list of objects is where we will eventually land for configuring the listeners in the `spec`. Eventual consistency, as it were ;-)

Reply:
Well, the Kube recommendation is to do exactly what we have there:

```yaml
listeners:
  plain:
    host: example.com
    port: 9092
  tls:
    host: example.com
    port: 9093
```

I have so far not seen any compelling use case for more interfaces. But the users are pretty much asking for having multiple interfaces for everything - not just for external - so it can be probably addressed only by a list:

```yaml
listeners:
  - type: plain
    host: example.com
    port: 9092
  - type: plain
    host: example.com
    port: 9093
  - type: plain
    host: example.com
    port: 9094
```

Reply:
Right, and my point is that if we need to use a list in the `spec` then we should be consistent and use a list in the `status`. Once we have this `status.listeners` people will start using it and it will be a breaking change to switch from using an object with properties, like

```
status:
  listeners:
    plain:
      # ...
    tls:
      # ...
```

to using a list like

```
status:
  listeners:
    - name: internal
      type: plain
      # ....
    - name: tls
      # ...
```

(It's a breaking change _anyway_, but while we can probably relax the schema to allow alternative representations in the `spec` users will care a lot more if we had to change the `status` which they've starting writing tooling/code around). 

And I think it's important that each listener has a unique name, so that we can use JSON path like

```
{.status.listeners[?(@.name==“plain”)].host}
```

to find a given listener. I know that's less convenient that having a named object, but I don't see an alternative.

Reply:
> {.status.listeners[?(@.name==“plain”)].host}

Did you write this out of your had? If yes you are good. But most people are not, so this is simply bad UX. As I said above - yes, if you expect us to add support for many listeners, we should use the list.